core/engine/src/value/inner/nan_boxed.rs

```diff
@@ -0,0 +1,864 @@
+//! This `[JsValue]` inner type is a NaN-boxed value, which is a 64-bits value
+//! that can represent any JavaScript value. If the integer is a non-NaN value,
+//! it will be stored as a 64-bits float. If it is a `f64::NAN` value, it will
+//! be stored as a quiet `NaN` value. Subnormal numbers are regular float.
+//!
+//! For any other type of values, the value will be stored as a 51-bits non-zero
+//! integer.
+//!
+//! In short, the memory layout of a NaN-boxed value is as follows:
+//!
+//! | Type of           | Bit Layout | Comment |
+//! |-------------------|------------|---------|
+//! | `+Infinity`       | `7FF0:0000:0000:0000`    | |
+//! | `-Infinity`       | `FFF0:0000:0000:0000`    | |
+//! | `NAN` (quiet)     | `7FF8:0000:0000:0000`    | |
+//! | `NAN` (signaling) | `FFF8:0000:0000:0000`    | |
```

Comment:
Why do we need a distinction here? When are they used separately?

Reply:
They are two different actual NaN values specced in IEEE 754; signaling just means that the CPU will raise an exception if used in any instruction. We should not mess with it, and it's something we can't use internally as it might lead to exceptions.